### PR TITLE
Rest api V2 - add GET environments/{envId}/apis/_search

### PR DESCRIPTION
**Pull request by ThibaudAV**
## Issue

https://gravitee.atlassian.net/browse/APIM-1163

## Description

Search apis with lucene

Work of the search : 
Step 0 : build apiQueryBuilder with value of ApiSearchQuery

Step 1 : find apiIds from lucene indexer from 'query' parameter without any pagination and sorting
Step 2 :(not linked to result of 1st step) if user is not admin, get list of apiIds in their scope and join with Lucene results
Step 3: get APIs page from repository by ids(intersection or Step 1 & Step 2 )  and add pagination and sorting

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-ustroqogql.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource.java [hunk 1]
```diff
@@ -77,4 +85,52 @@ public ApisResponse getApis(@BeanParam @Valid PaginationParam paginationParam) {
...
+        apiSortByParam.validate();
+
+        QueryBuilder<ApiEntity> apiQueryBuilder = QueryBuilder.create(ApiEntity.class);
+
+        if (!Strings.isNullOrEmpty(apiSearchQuery.getQuery())) {
+            apiQueryBuilder.setQuery(apiSearchQuery.getQuery());
+        }
+
+        if (Objects.nonNull(apiSearchQuery.getIds()) && !apiSearchQuery.getIds().isEmpty()) {
+            apiQueryBuilder.addFilter(FIELD_TYPE_VALUE, apiSearchQuery.getIds());
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource.java]:
```suggestion
            apiQueryBuilder.addFilter(FIELD_ID, apiSearchQuery.getIds());
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource_SearchApisTest.java [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        assertNotNull(apiQuery.getFilters().get("api"));
+        List<String> ids = (List<String>) apiQuery.getFilters().get("api");
+        assertEquals(2, ids.size());
+        assertEquals("id-2", ids.get(1));
+
+        assert (!apiQuery.getFilters().containsKey("definition_version"));
+    }
+
+    @Test
+    public void should_return_error_with_wrong_value_for_order_by_param() {
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource_SearchApisTest.java]:
```suggestion
    public void should_return_error_with_wrong_value_for_sort_by_param() {
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource_SearchApisTest.java [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+
+        final Response response = rootTarget().queryParam("sortBy", "wrong_value").request().post(Entity.json(apiSearchQuery));
+        assertEquals(HttpStatusCode.BAD_REQUEST_400, response.getStatus());
+
+        Map<String, Object> body = response.readEntity(Map.class);
+        assert (String.valueOf(body.get("message")).endsWith("wrong_value"));
+    }
+
+    @Test
+    public void should_sort_results_with_order_by_param_asc() {
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource_SearchApisTest.java]:
```suggestion
    public void should_sort_results_with_sort_by_param_asc() {
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource_SearchApisTest.java [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+        assertEquals("api-id", data.get(0).getApiV4().getId());
+
+        var apiQueryBuilder = apiQueryBuilderCaptor.getValue();
+        var apiQuery = apiQueryBuilder.build();
+        assertEquals("api-name", apiQuery.getQuery());
+        assertEquals(0, apiQuery.getFilters().size());
+    }
+
+    @Test
+    public void should_sort_results_with_order_by_param_desc() {
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource_SearchApisTest.java]:
```suggestion
    public void should_sort_results_with_sort_by_param_desc() {
```

**Review by phiz71 — CHANGES_REQUESTED** [review]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource.java (hunk 1 above)
Comment (ThibaudAV) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource.java]: hahaha If only it were that simple . but no without refactor it is well api for a dark unknown reason

Comment (phiz71) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApisResource.java]: Ok, I've looked into the code... it's a bit messy

**Review by phiz71 — APPROVED** [review]

**Pull request merged**